File: app/src/main/java/de/nils_beyer/android/Vertretungen/mainActivity/OverviewAdapter.java
package de.nils_beyer.android.Vertretungen.mainActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.io.Serializable;
import java.text.SimpleDateFormat;

import de.nils_beyer.android.Vertretungen.R;
import de.nils_beyer.android.Vertretungen.data.Group;
import de.nils_beyer.android.Vertretungen.data.GroupCollection;
import de.nils_beyer.android.Vertretungen.data.TeacherGroup;
import de.nils_beyer.android.Vertretungen.detailActivity.DetailActivity;
import de.nils_beyer.android.Vertretungen.preferences.MarkedKlasses;
import de.nils_beyer.android.Vertretungen.util.DateParser;

public class OverviewAdapter extends RecyclerView.Adapter<OverviewAdapter.ViewHolder> {
    private Context context;
    private GroupCollection groupCollection;

    public OverviewAdapter(Context c, GroupCollection groupCollection) {
        context = c;
        this.groupCollection = groupCollection;
    }


    class ViewHolder extends RecyclerView.ViewHolder {
        TextView className;
        TextView replacementCounter;
        CardView cardView;
        ImageView marked;


        ViewHolder(View v) {
            super(v);
            cardView = (CardView) v.findViewById(R.id.overview_card);
            className = (TextView) v.findViewById(R.id.text_class_name);
            replacementCounter = (TextView) v.findViewById(R.id.text_replacement_count);
            marked = (ImageView) v.findViewById(R.id.image_class_marked);

        }

        void bind(final int position) {
            Group group = groupCollection.getGroupArrayList().get(position);
            if (group instanceof TeacherGroup) {
                className.setText(String.format(context.getString(R.string.overview_adapter_group_teacher), group.name));
            } else {
                className.setText(String.format(context.getString(R.string.overview_adapter_group_student), group.name));
            }

            if (group.replacements.size() > 1)
                replacementCounter.setText(String.format(context.getString(R.string.overview_adapter_entry_multiple), String.valueOf(group.replacements.size())));
            else
                replacementCounter.setText(String.format(context.getString(R.string.overview_adapter_entry_single), String.valueOf(group.replacements.size())));

            marked.setVisibility(group.isMarked(context) ? View.VISIBLE : View.GONE);

            cardView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    DetailActivity.startActivity(context, groupCollection, position);
                }
            });
        }
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        if (position == 0) {
            holder.cardView.setOnClickListener(null);
            holder.replacementCounter.setVisibility(View.GONE);
            holder.className.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
            holder.cardView.setClickable(false);
            holder.marked.setVisibility(View.GONE);
            if (groupCollection.getImmediacity() != null) {
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat(context.getString(R.string.date_format_only_time));
                holder.className.setText(String.format(
                        context.getString(R.string.overview_adapter_immediacity),
                        DateParser.parseDateToShortString(context, groupCollection.getImmediacity()),
                        simpleDateFormat.format(groupCollection.getImmediacity()
                )));
            }
            else {
                holder.cardView.setVisibility(View.GONE);
            }
        } else {
            holder.cardView.setClickable(true);
            holder.cardView.setVisibility(View.VISIBLE);
            holder.replacementCounter.setVisibility(View.VISIBLE);
            holder.marked.setVisibility(View.VISIBLE);
            holder.className.setTextAlignment(View.TEXT_ALIGNMENT_TEXT_START);
            holder.bind(position - 1);


        }

    }

    @Override
    public int getItemCount() {
        if (groupCollection.getGroupArrayList().size() == 0) {
            return 0;
        } else {
            return groupCollection.getGroupArrayList().size() + 1;
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.recyclerview_overview_layout, parent, false);
        ViewHolder vh = new ViewHolder(v);
        return vh;
    }

    public void update(@Nullable GroupCollection collection) {
        if (collection == null) {
            return;
        }
        this.groupCollection = collection;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/de/nils_beyer/android/Vertretungen/mainActivity/OverviewAdapter.kt
package de.nils_beyer.android.Vertretungen.mainActivity

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup

import de.nils_beyer.android.Vertretungen.R
import de.nils_beyer.android.Vertretungen.account.Account
import de.nils_beyer.android.Vertretungen.account.Dataset
import de.nils_beyer.android.Vertretungen.data.GroupCollection
import de.nils_beyer.android.Vertretungen.events.Event
import de.nils_beyer.android.Vertretungen.events.EventStorage
import de.nils_beyer.android.Vertretungen.mainActivity.viewholder.EventViewHolder
import de.nils_beyer.android.Vertretungen.mainActivity.viewholder.GroupViewHolder
import de.nils_beyer.android.Vertretungen.mainActivity.viewholder.ImmediacityViewHolder
import de.nils_beyer.android.Vertretungen.mainActivity.viewholder.MOTDHolder

class OverviewAdapter(private val context: Context, private var groupCollection: GroupCollection, private val account : Dataset) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    enum class ViewTypes { GroupViewHolder, ImmediacityViewHolder, EventsViewHolder, MOTDHolder }

    private var events : List<Event> = loadEvents()
    private var motd : String? = account.getMOTD(context)



    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (motd == null) {
            when (holder) {
                is GroupViewHolder -> holder.bind(groupCollection, position - events.size - 1)
                is ImmediacityViewHolder -> holder.bind(groupCollection.immediacity)
                is EventViewHolder -> holder.bind(events[position])
            }
        } else {
            when (holder) {
                is GroupViewHolder -> holder.bind(groupCollection, position - events.size)
                is ImmediacityViewHolder -> holder.bind(groupCollection.immediacity)
                is EventViewHolder -> holder.bind(events[position - 1])
                is MOTDHolder -> holder.bind(motd!!)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (motd == null) {
            val eventsCount = events.size
            return when {
                position < eventsCount -> ViewTypes.EventsViewHolder.ordinal
                position == eventsCount -> ViewTypes.ImmediacityViewHolder.ordinal
                else -> ViewTypes.GroupViewHolder.ordinal
            }
        } else {
            val eventsCount = events.size
            return when {
                position == 0 -> ViewTypes.MOTDHolder.ordinal
                position < eventsCount + 1 -> ViewTypes.EventsViewHolder.ordinal
                position == eventsCount + 1 -> ViewTypes.ImmediacityViewHolder.ordinal
                else -> ViewTypes.GroupViewHolder.ordinal
            }
        }
    }

    override fun getItemCount(): Int {
        var count = events.size
        if (groupCollection.groupArrayList.size != 0) {
            count += groupCollection.groupArrayList.size + 1
        }
        if (motd != null) {
            Log.d("MOTD", "MOTD fount: " + motd)
            count += 1
        }
        return count
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            ViewTypes.ImmediacityViewHolder.ordinal -> {
                val v = LayoutInflater.from(parent.context)
                        .inflate(R.layout.recyclerview_overview_immediacity, parent, false)
                return ImmediacityViewHolder(v, context)
            }
            ViewTypes.GroupViewHolder.ordinal -> {
                val v = LayoutInflater.from(parent.context)
                        .inflate(R.layout.recyclerview_overview_layout, parent, false)
                return GroupViewHolder(v, context)
            }
            ViewTypes.EventsViewHolder.ordinal -> {
                val v = LayoutInflater.from(parent.context)
                        .inflate(R.layout.recyclerview_overview_event, parent, false)
                return EventViewHolder(v, context)
            }
            ViewTypes.MOTDHolder.ordinal -> {
                val v = LayoutInflater.from(parent.context)
                        .inflate(R.layout.recyclerview_overview_motd, parent, false)
                return MOTDHolder(v, context)
            }
            else -> TODO("not yet implemented?!")
        }
    }

    fun update(collection: GroupCollection?) {
        if (collection == null) {
            return
        }
        this.groupCollection = collection
        this.motd = account.getMOTD(context)
        loadEvents()

        notifyDataSetChanged()
    }

    private fun loadEvents() : List<Event> {
        // collection.date can be null, for instance when using EmptyAccount (no account is registered!)
        // TODO Handle this better. Remove empty account
        if (groupCollection.date != null) {
            this.events = EventStorage.getEventsAt(context, groupCollection.date)
        } else {
            this.events = ArrayList<Event>()
        }
        return this.events
    }
}